keras/src/ops/core.py

```diff
@@ -199,6 +201,97 @@ def scan(f, init, xs, length=None):
     )
 
 
+class AssociativeScan(Operation):
+    def __init__(self, reverse=False):
+        super().__init__()
+        self.reverse = reverse
+
+    def call(self, f, elems, axis=0):
+        return backend.core.associative_scan(
+            f, elems, reverse=self.reverse, axis=axis
+        )
+
+    def compute_output_spec(self, f, elems, axis):
+        elems_flat, tree_ = optree.tree_flatten(elems)
+        lens = [elem.shape[axis] for elem in elems_flat]
+        if len(set(lens)) != 1:
+            raise ValueError(
+                "Array inputs to associative_scan must have the same "
+                "first dimension. (saw: {})".format(
+                    [elem.shape for elem in elems_flat]
+                )
+            )
+
+        x = optree.tree_unflatten(
+            tree_, [slice_along_axis(x, 0, 1, axis=axis) for x in elems_flat]
+        )
+        y_spec = backend.compute_output_spec(f, x, x)
+
+        def _restore_shape(x):
+            return KerasTensor(
+                shape=elems_flat[0].shape, dtype=x.dtype, sparse=x.sparse
+            )
+
+        y_spec = tree.map_structure(_restore_shape, y_spec)
+        return y_spec
+
+
+@keras_export("keras.ops.associative_scan")
+def associative_scan(f, elems, reverse=False, axis=0):
+    """Performs a scan with an associative binary operation, in parallel.
```

Comment:
Below, it would be worth introducing more info about use cases for this op. The docstring should answer, what does this do, and why would I want to use it? Perhaps contrast it with `scan`.